clickhouse-core/src/main/scala/xenon/clickhouse/hash/HashFunc.scala

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xenon.clickhouse.hash
+
+import scala.reflect.ClassTag
+
+abstract class HashFunc[T: ClassTag] {
+  def applyHash(input: Array[Byte]): T
+  def combineHashes(h1: T, h2: T): T
+
+  final def executeAny(input: Any): T =
+    input match {
+      case input: String => applyHash(input.getBytes)
```

Comment:
```suggestion
      case bytes: Array[Byte] => applyHash(bytes)
      case input: String => applyHash(input.getBytes(StandardCharsets.UTF_8))
```

Reply:
`applyHash(string.getBytes(StandardCharsets.UTF_8))` is needed, and I forgot the encoding issue. 

For `Array[Byte]`, I have a concern is that in Clickhouse, we don't have byte array, but we have `Array[UInt8]`. [Array type is handled differently](https://github.com/ClickHouse/ClickHouse/blob/v23.5.3.24-stable/src/Functions/FunctionsHashing.h#L1265) from byte array. In the future if we support `Array[UInt8]`, we cannot tell it from byte array. 

I suppose the reason for adding `Array[Byte]` is e.g. we are using spark's `UTF8String` when we actually call it, and `UTF8String.getBytes` is better than `UTF8String.toString`(for cost of reconstructing String and so on). I think this is a trade-off between usage and better supporting Clickhouse's Array type. My original view is that, when we need to support one more type of Clickhouse, we add that type in `executeAny `. If we want to stick to this patch, at least we need to comment this issue in the code.

Reply:
> `UTF8String.getBytes` is better than `UTF8String.toString`

That's right.

> If we want to stick to this patch, at least we need to comment this issue in the code.

OK, let's do it.

Reply:
Updated as discussed